src/jit/optimizer.cpp

```diff
@@ -7003,22 +7015,32 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                 // Is the value of the whole tree loop invariant?
                 if (!treeIsInvariant)
                 {
+                    // Here we have a tree that is not loop invariant and we thus cannot hoist
                     treeIsHoistable = false;
+
+                    // Check if we should also clear m_beforeSideEffect.
+                    if (m_beforeSideEffect)
+                    {
+                        // If 'tree' can throw an exception then we need to set m_beforeSideEffect to false.
+                        if (tree->OperMayThrow(m_compiler))
```

Comment:
This call only checks is the operation itself may throw; it doesn't check whether the children may throw. I think the right check is `if ((tree->gtFlags & GTF_EXCEPT) != 0)`.

Also, I think this check belongs to the block below starting with https://github.com/dotnet/coreclr/pull/26952/files#diff-15cd21de3f1ffaa677e7da145576d60cL7013-L7016
(where we also checks for calls and assignments) so that it's done regardless of `treeIsInvariant` value.

Reply:
I used to have that check, but I refined it to this one.  

When we get here we currently have `m_beforeSideEffect `set to true, and have a tree whose current node is no longer invariant, but it's children were invariant and so are ready to be hoisted out of the loop.

With the old check we would observe any`GTF_EXCEPT ` flag bubbling up from the child nodes, and thus would clear `m_beforeSideEffect`, thus prevent us from doing any further hoisting, even though we haven't moved an expression across another node with an exception side effect.

Reply:
Related test mentioned in #26865

Reply:
Looks like my current change also regresses  JI\Regression\CLR-x86-JIT\V1-M12-Beta2\b79250\b79250\b79250.cmd (x86 checked)